sdk/search/Azure.Search.Documents/tests/Serialization/SearchDocumentTests.cs

```diff
@@ -0,0 +1,761 @@
+                    $"[{new ComplexNullable(1, true, "hi").ToJson()}, {new ComplexNullable(2, false, "bye").ToJson()}]",
+                    new[] { new ComplexNullable(1, true, "hi").ToDocument(), new ComplexNullable(2, false, "bye").ToDocument() }),
+                TestValue<SearchDocument[]>.Exact(
+                    $"[{new Complex().ToJson()}, {new ComplexNullable().ToJson()}]",
+                    new[] { new Complex().ToDocument(), new ComplexNullable().ToDocument() }));
+
+#if EXPERIMENTAL_DYNAMIC
+        [Test]
+#endif
+        public void Index()
+        {
+            dynamic doc = ToDocument("12");
+            int value = doc.Value;
+            SearchTestBase.AssertApproximate(12, value);
+        }
+
+#if EXPERIMENTAL_DYNAMIC
+        [Test]
+#endif
+        public void IndexComplex()
+        {
+            dynamic doc = ToDocument(new Complex(1, true, "hi").ToJson());
+
+            int a = doc.Value.A;
+            SearchTestBase.AssertApproximate(1, a);
+
+            bool b = doc.Value.B;
+            SearchTestBase.AssertApproximate(true, b);
+
+            string c = doc.Value.C;
+            SearchTestBase.AssertApproximate("hi", c);
+        }
+
+#if EXPERIMENTAL_DYNAMIC
+        [Test]
+#endif
+        public void IndexNestedComplex()
+        {
+            dynamic doc = ToDocument(new NestedComplex(1, true, "hi").ToJson());
+
+            int a = doc.Value.Nested.A;
+            SearchTestBase.AssertApproximate(1, a);
+
+            bool b = doc.Value.Nested.B;
+            SearchTestBase.AssertApproximate(true, b);
+
+            string c = doc.Value.Nested.C;
+            SearchTestBase.AssertApproximate("hi", c);
+        }
+
+        [Test]
+        public void Equality()
+        {
+            SearchTestBase.AssertApproximate(
+                new SearchDocument(),
+                new SearchDocument());
+
+            Assert.AreNotEqual(
+                new SearchDocument(),
+                new Complex().ToDocument());
+            SearchTestBase.AssertApproximate(
+                new Complex().ToDocument(),
+                new Complex().ToDocument());
+
+            Assert.AreNotEqual(
+                new Complex(1, false, "hi").ToDocument(),
+                new NestedComplex(1, false, "hi").ToDocument());
+            SearchTestBase.AssertApproximate(
+                new NestedComplex(1, false, "hi").ToDocument(),
+                new NestedComplex(1, false, "hi").ToDocument());
+            SearchTestBase.AssertApproximate(
+                new Complex(1, false, "hi").ToDocument(),
+                new NestedComplex(1, false, "hi").Nested.ToDocument());
+        }
+
+#if EXPERIMENTAL_DYNAMIC
+        [Test]
+#endif
+        public void NoDynamicObjectMembers()
+        {
+            Assert.IsInstanceOf<IDynamicMetaObjectProvider>(new SearchDocument());
+            Assert.IsNotInstanceOf<DynamicObject>(new SearchDocument());
+        }
+    }
+}
```

Comment:
Probably should test recursion as well (make sure we throw in the cases we expect). Be sure to include some collection properties, though - it looks like how the code is now, recursive depth is decremented incorrectly and you may throw prematurely (i.e. test right up to the max specified, then 1 or more past it).